src/x11.cc

```diff
@@ -1139,44 +1141,46 @@ void set_struts(alignment align) {
+    int display_height = workarea.height();
 
     switch (horizontal_alignment(align)) {
       case axis_align::START:
         sizes[*x11_strut::LEFT] = std::clamp(
-            window.geometry.x + window.geometry.width, 0, *workarea.width);
+            window.geometry.x() + window.geometry.end_x(), 0, display_width);
         sizes[*x11_strut::LEFT_START_Y] =
-            std::clamp(*window.geometry.y, 0, *workarea.height);
+            std::clamp(window.geometry.y(), 0, display_height);
         sizes[*x11_strut::LEFT_END_Y] = std::clamp(
-            window.geometry.y + window.geometry.height, 0, *workarea.height);
+            window.geometry.y() + window.geometry.height(), 0, display_height);
         break;
       case axis_align::END:
         sizes[*x11_strut::RIGHT] =
-            std::clamp(workarea.width - window.geometry.x, 0, *workarea.width);
+            std::clamp(display_width - window.geometry.x(), 0, display_width);
         sizes[*x11_strut::RIGHT_START_Y] =
-            std::clamp(*window.geometry.y, 0, *workarea.height);
+            std::clamp(window.geometry.y(), 0, display_height);
         sizes[*x11_strut::RIGHT_END_Y] = std::clamp(
-            window.geometry.y + window.geometry.height, 0, *workarea.height);
+            window.geometry.y() + window.geometry.height(), 0, display_height);
         break;
       case axis_align::MIDDLE:
         switch (vertical_alignment(align)) {
           case axis_align::START:
             sizes[*x11_strut::TOP] =
-                std::clamp(window.geometry.y + window.geometry.height, 0,
-                           *workarea.height);
+                std::clamp(window.geometry.y() + window.geometry.height(), 0,
+                           display_height);
             sizes[*x11_strut::TOP_START_X] =
-                std::clamp(*window.geometry.x, 0, *workarea.width);
-            sizes[*x11_strut::TOP_END_X] = std::clamp(
-                window.geometry.x + window.geometry.width, 0, *workarea.width);
+                std::clamp(window.geometry.x(), 0, display_width);
+            sizes[*x11_strut::TOP_END_X] =
+                std::clamp(window.geometry.x() + window.geometry.width(), 0,
+                           display_width);
             break;
           case axis_align::END:
             sizes[*x11_strut::BOTTOM] = std::clamp(
-                workarea.height - window.geometry.y, 0, *workarea.height);
+                display_height - window.geometry.y(), 0, display_height);
             sizes[*x11_strut::BOTTOM_START_X] =
-                std::clamp(*window.geometry.x, 0, *workarea.width);
-            sizes[*x11_strut::BOTTOM_END_X] = std::clamp(
-                window.geometry.x + window.geometry.width, 0, *workarea.width);
+                std::clamp(window.geometry.x(), 0, display_width);
+            sizes[*x11_strut::BOTTOM_END_X] =
+                std::clamp(window.geometry.x() + window.geometry.width(), 0,
+                           display_width);
```

Comment:
This was buggy. I didn't use `display_width` and `display_height` at all, and instead `workarea.width` which was in fact an absolute end x position and not width.